queries/c/highlights.scm

```diff
@@ -191,6 +190,9 @@
 (parameter_declaration
   declarator: (pointer_declarator) @parameter)
 
+(parameter_declaration
+  declarator: (pointer_declarator (identifier) @parameter))
```

Comment:
the problem with this is that pointers can nest arbitrarily. `*foo` `**foo` `***foo` `&foo`, `&&foo`.

`has_ancestor` was introduced for these and we had queries until nesting level 2 before. We successfully removed that costly operation by lowering the priority of `@variable`. I wouldn't want to change that just for issues with injections.

I think this is rather general issues with injections where there some situation where you want to have the injection having priority over everything in the parent language. Do we need `literal` highlighting in markdown when there's an injection. Probably the `literal` highlighting should have a very low priority when it's meant that injections render over it.

Reply:
`& foo` and all variations thereof already render correctly due to precedence; it's just `* foo` (and higher orders) that don't.

I would prefer to keep modifications within the language, if possible (since this is a C-specific issue; other injections have no problem).

Otherwise your suggestion works, too; we just need to make sure to keep high enough priority to avoid unwanted issues with other highlights (legacy syntax if enabled etc.)

Reply:
I tried raising the priority of the pointer declarator instead; this seems to highlight every variant your mention correctly.

Reply:
Wouldn't that require to also raise the priority of the `*` operator? The idea of lowering variable vs increasing pointer declarator was that the operators would still paint over the actual to large declarator range. But if raising operators too would solve the problem we can of course do it.

In general, I'd argue that this isn't a C issue but a markdown one. It will happen for all code injections with languages with different priorities. Similar issues can also happen when injecting C into C macros.

In general, `(identifier) @variable` is so unspecific that it will also compete with other highlights and would then again just get it's priority over type identifiers (etc) only via ordering.

Reply:
> Wouldn't that require to also raise the priority of the * operator? The idea of lowering variable vs increasing pointer declarator was that the operators would still paint over the actual to large declarator range. But if raising operators too would solve the problem we can of course do it.

I didn't see an issue with the operators; the original issue was only about the arguments themselves?

(I can of course do that, too, but I'd prefer to only make changes that are _necessary_ for the issue at hand.)

> In general, I'd argue that this isn't a C issue but a markdown one. It will happen for all code injections with languages with different priorities. Similar issues can also happen when injecting C into C macros.

Actually, the issue is even more general, with _any_ injection into a node that also has highlights defined (as you point out). So if we want to solve this on that level (which is a perfectly reasonable idea), we need a more general solution; preferably one that doesn't involve having to reason about all possible combinations.

(This is what I'm worried about here: having to keep in mind all possible priorities of the injected language. So there needs to be one(!) priority for injection targets that is the same and works for all languages, so we can document it.)

> In general, (identifier) @variable is so unspecific that it will also compete with other highlights and would then again just get it's priority over type identifiers (etc) only via ordering.

Yes, but that is precisely what precedence by ordering is for?

Reply:
> (I can of course do that, too, but I'd prefer to only make changes that are necessary for the issue at hand.)

I agree, but `@variable` was never really supported by nvim-treesitter (was treated as unset highlight). It's possible that it has more problems on a global level and language injections in markdown seem like a side-issue when compared to issues that might be still in the actual languages. It could wait until we revise `@variable` globally and see that is works well and similar in most languages.

> I didn't see an issue with the operators; the original issue was only about the arguments themselves?

The solution with highlighting `pointer_declarator` as parameter has the problem that it will hightlight all `int* foo` as parameter instead of just `foo` which relies on `*` and `int` overwriting the highlight. So, we have to be a bit careful when raising prio of `pointer_declarator`-

> Yes, but that is precisely what precedence by ordering is for?

If it works, I'm fine with the change. We just need to do a smoke test that everything still works as intended.

Reply:
>  It's possible that it has more problems on a global level and language injections in markdown seem like a side-issue when compared to issues that might be still in the actual languages.

Well, not side-issue, exactly, but I agree that these are two different but interacting issues. Let's proceed with your suggestion then (lowering the priority of code blocks with possible injections).

Reply:
increasing the prio of `@variable` for C (or just make it the same for all languages) could help to be more inline with other languages. We should try to apply both measures long term (not in a single PR)